### PR TITLE
2018 11 21 performance improvements

### PR DESCRIPTION
**Pull request by Christewart**
This seems to improve performance by roughly 2x with rudimentary bench markets 

```
2018-11-21 16:28:52.323 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 1593ms
2018-11-21 16:28:53.052 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 712ms
2018-11-21 16:28:53.729 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 648ms
2018-11-21 16:28:54.072 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 328ms
2018-11-21 16:28:54.459 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 372ms
2018-11-21 16:28:54.793 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 320ms
2018-11-21 16:28:55.200 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 394ms
2018-11-21 16:28:55.528 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 315ms
2018-11-21 16:28:56.002 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 460ms
2018-11-21 16:28:56.407 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 388ms

Flat profile of 5.97 secs (533 total ticks): main

  Interpreted + native   Method                        
  0.9%     4  +     1    java.lang.ClassLoader.defineClass1
  0.6%     3  +     0    java.lang.Integer.toString
  0.6%     3  +     0    org.bitcoins.core.util.BitcoinScriptUtil$class.castToBool
  0.6%     3  +     0    java.math.BigInteger.toByteArray
  0.4%     0  +     2    java.util.zip.ZipFile.getEntryTime
  0.4%     0  +     2    org.bitcoin.NativeSecp256k1.secp256k1_ec_pubkey_decompress
  0.4%     2  +     0    org.bitcoins.core.script.reserved.UndefinedOP_NOP.bytes
  0.4%     2  +     0    org.bitcoins.core.script.constant.StackPushOperationFactory$class.operations
  0.2%     0  +     1    java.lang.Class.isInterface
  0.2%     0  +     1    org.bitcoin.Secp256k1Context.secp256k1_init_context
  0.2%     1  +     0    java.util.zip.Inflater.reset
  0.2%     0  +     1    java.security.AccessController.doPrivileged
  0.2%     1  +     0    org.bitcoins.core.protocol.script.ScriptPubKey$.fromAsm
  0.2%     1  +     0    org.bitcoins.core.number.Int64$$anonfun$fromBytes$4.<init>
  0.2%     1  +     0    org.bitcoins.core.script.stack.OP_3DUP$.bytes
  0.2%     1  +     0    org.bitcoins.core.script.stack.OP_2ROT$.bytes
  0.2%     1  +     0    org.bitcoins.core.script.stack.OP_SWAP$.bytes
  0.2%     1  +     0    org.bitcoins.core.script.stack.OP_TUCK$.bytes
  0.2%     1  +     0    ch.qos.logback.core.subst.Parser.makeNewLiteralNode
  0.2%     1  +     0    org.bitcoins.core.crypto.ECPublicKey$.fromBytes
  0.2%     1  +     0    org.bitcoins.core.protocol.script.P2PKHScriptSignature$$anonfun$fromAsm$4.apply
  0.2%     1  +     0    org.bitcoins.core.crypto.DoubleSha256Digest$.fromBytes
  0.2%     1  +     0    scodec.bits.ByteVector.toArray
  0.2%     1  +     0    scala.LowPriorityImplicits.wrapByteArray
  0.2%     1  +     0    org.bitcoins.core.script.stack.OP_IFDUP$.<init>
 12.8%    59  +     9    Total interpreted (including elided)

     Compiled + native   Method                        
  3.2%    13  +     4    scodec.bits.ByteVector$.fromHexInternal
  2.3%    12  +     0    scala.collection.IndexedSeqOptimized$class.foreach
  2.1%    11  +     0    org.bitcoins.core.serializers.script.ScriptParser.loop$2
  1.9%    10  +     0    org.bitcoins.core.util.BitcoinSUtil$class.encodeHex
  1.7%     9  +     0    scodec.bits.ByteVector.go$5
  1.7%     9  +     0    java.math.BigInteger.<init>
  1.5%     7  +     1    scala.collection.TraversableOnce$$anonfun$addString$1.apply
  1.3%     7  +     0    scodec.bits.ByteVector$Buffer.$plus$plus
  1.1%     6  +     0    org.bitcoins.core.protocol.script.MultiSignatureScriptPubKey$.isMultiSignatureScriptPubKey
  0.9%     5  +     0    scala.collection.AbstractIterable.foreach
  0.9%     0  +     5    scala.reflect.ManifestFactory$$anon$6.newArray
  0.9%     5  +     0    org.bitcoins.core.script.constant.ScriptNumberOperation.toByte
  0.9%     5  +     0    org.bitcoins.core.util.BitcoinSUtil$$anonfun$toByteVector$1.apply
  0.9%     5  +     0    scala.collection.TraversableLike$class.$plus$plus
  0.8%     4  +     0    org.bitcoins.core.protocol.script.MultiSignatureScriptPubKey$.org$bitcoins$core$protocol$script$MultiSignatureScriptPubKey$$isValidPubKeyNumber
  0.8%     3  +     1    org.bitcoins.core.serializers.RawSerializerHelper.writeCmpctSizeUInt
  0.8%     4  +     0    org.bitcoins.core.protocol.CompactSizeUInt.hex
  0.6%     3  +     0    org.bitcoins.core.script.constant.ScriptNumberOperation.bytes
  0.6%     3  +     0    scala.collection.mutable.StringBuilder.appendAll
  0.6%     2  +     1    scala.collection.immutable.List.foreach
  0.6%     2  +     1    org.bitcoins.core.script.constant.ScriptNumberUtil$class.toLong
  0.4%     2  +     0    scala.collection.immutable.List.find
  0.4%     2  +     0    scodec.bits.ByteVector$Buffer.getImpl
  0.4%     2  +     0    scala.runtime.ScalaRunTime$._toString
  0.4%     2  +     0    org.bitcoins.core.protocol.script.ScriptFactory$class.buildScript
 40.0%   176  +    37    Total compiled (including elided)

         Stub + native   Method                        
 27.8%     0  +   148    org.bitcoin.NativeSecp256k1.secp256k1_ec_pubkey_decompress
  6.0%     0  +    32    java.lang.Throwable.fillInStackTrace
  3.8%     0  +    20    java.lang.reflect.Array.newArray
  3.4%     0  +    18    java.lang.Object.clone
  2.3%     0  +    12    sun.misc.Unsafe.copyMemory
  1.9%     0  +    10    java.util.zip.Inflater.inflateBytes
  1.1%     5  +     1    java.lang.ClassLoader.defineClass1
  0.2%     0  +     1    java.lang.Class.isInterface
  0.2%     0  +     1    java.security.AccessController.doPrivileged
  0.2%     0  +     1    java.lang.ClassLoader.findBootstrapClass
  0.2%     0  +     1    java.lang.ClassLoader.findLoadedClass0
  0.2%     0  +     1    java.lang.System.arraycopy
  0.2%     0  +     1    java.util.zip.ZipFile.getEntry
 47.3%     5  +   247    Total stub


Global summary of 5.99 seconds:
100.0%   542             Received ticks
  1.3%     7             Received GC ticks
142.8%   774             Compilation

```

### CODE REVIEW

**Comment by Christewart** [comment]
Latest profile on c4b665be1c387b0930b4a5d12bb8943c4775844e

```
2018-11-22 11:33:41.103 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 995ms
2018-11-22 11:33:41.534 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 416ms
2018-11-22 11:33:42.035 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 485ms
2018-11-22 11:33:42.333 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 284ms
2018-11-22 11:33:42.728 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 383ms
2018-11-22 11:33:43.003 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 263ms
2018-11-22 11:33:43.267 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 252ms
2018-11-22 11:33:43.577 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 295ms
2018-11-22 11:33:43.846 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 257ms
2018-11-22 11:33:44.179 INFO  BlockBench$.timeBlockParsing(14) - Elapsed time: 320ms

Flat profile of 4.35 secs (397 total ticks): main

  Interpreted + native   Method                        
  1.3%     5  +     0    scala.collection.GenIterable$class.$init$
  1.3%     5  +     0    java.lang.ClassLoader.defineClass1
  0.3%     0  +     1    org.bitcoin.Secp256k1Context.secp256k1_init_context
  0.3%     1  +     0    scodec.bits.ByteVector.tail
  0.3%     0  +     1    java.lang.Throwable.fillInStackTrace
  0.3%     1  +     0    org.bitcoins.core.protocol.script.ScriptPubKey$.apply
  0.3%     1  +     0    scala.collection.immutable.Range.isEmpty
  0.3%     1  +     0    org.bitcoins.core.script.splice.OP_CAT$.opCode
  0.3%     0  +     1    sun.misc.Unsafe.copyMemory
  0.3%     1  +     0    java.lang.Character.isWhitespace
  0.3%     1  +     0    scala.collection.immutable.Range$.apply
  0.3%     1  +     0    scala.runtime.ScalaRunTime$$anon$1.<init>
  0.3%     1  +     0    scala.App$class.$init$
  0.3%     1  +     0    org.bitcoins.core.protocol.script.P2WPKHWitnessV0$.apply
  0.3%     1  +     0    scala.collection.immutable.VectorPointer$class.copyRange
  0.3%     1  +     0    org.bitcoins.core.protocol.script.P2WPKHWitnessSPKV0$.isValid
  0.3%     1  +     0    scodec.bits.ByteVector.bufferBy
  0.3%     1  +     0    org.bitcoins.core.protocol.script.ScriptFactory$class.buildScript
  0.3%     1  +     0    org.bitcoins.core.protocol.script.P2WPKHWitnessSPKV0$.fromAsm
  0.3%     0  +     1    ch.qos.logback.core.joran.GenericConfigurator.doConfigure
  0.3%     1  +     0    java.util.zip.ZipInputStream.<init>
  0.3%     1  +     0    sun.reflect.generics.parser.SignatureParser.parseClassTypeSignature
  0.3%     0  +     1    ch.qos.logback.classic.LoggerContext.<init>
  0.3%     1  +     0    org.bitcoins.core.serializers.script.RawScriptWitnessParser.loop$1
  0.3%     1  +     0    org.bitcoins.core.serializers.transaction.RawWitnessTransactionParser.write
 10.6%    37  +     5    Total interpreted (including elided)

     Compiled + native   Method                        
  3.3%    11  +     2    scodec.bits.ByteVector$.fromHexInternal
  2.3%     9  +     0    scala.collection.IndexedSeqOptimized$class.foreach
  2.3%     9  +     0    org.bitcoins.core.serializers.script.ScriptParser.loop$2
  1.5%     0  +     6    scala.reflect.ManifestFactory$$anon$6.newArray
  1.5%     6  +     0    org.bitcoins.core.script.constant.ScriptNumberOperation.toByte
  1.3%     5  +     0    org.bitcoins.core.script.arithmetic.ArithmeticOperation.toByte
  1.3%     5  +     0    scala.collection.TraversableOnce$$anonfun$addString$1.apply
  1.0%     4  +     0    org.bitcoins.core.protocol.script.MultiSignatureScriptSignature$$anonfun$11.apply
  1.0%     4  +     0    org.bitcoins.core.number.UInt64.encodeHex
  1.0%     4  +     0    scodec.bits.ByteVector.toString
  1.0%     4  +     0    scodec.bits.ByteVector.go$5
  1.0%     3  +     1    java.math.BigInteger.<init>
  0.8%     3  +     0    org.bitcoins.core.serializers.transaction.RawTransactionInputParser.read
  0.8%     3  +     0    org.bitcoins.core.protocol.script.MultiSignatureScriptPubKey$.isMultiSignatureScriptPubKey
  0.5%     2  +     0    ExceptionBlob
  0.5%     2  +     0    scala.collection.AbstractIterable.foreach
  0.5%     2  +     0    org.bitcoins.core.number.UInt32.hex
  0.5%     1  +     1    scala.Predef$.require
  0.5%     2  +     0    org.bitcoins.core.number.UInt32$.fromBytes
  0.5%     2  +     0    scala.collection.TraversableLike$class.$plus$plus
  0.5%     1  +     1    scala.collection.IndexedSeqOptimized$class.zipWithIndex
  0.5%     2  +     0    org.bitcoins.core.serializers.transaction.RawBaseTransactionParser.write
  0.5%     2  +     0    org.bitcoins.core.util.BitcoinSUtil$class.encodeHex
  0.5%     2  +     0    org.bitcoins.core.protocol.script.ScriptPubKey$.fromAsm
  0.3%     1  +     0    scala.collection.immutable.List.find
 32.2%   107  +    21    Total compiled (including elided)

         Stub + native   Method                        
 35.0%     0  +   139    org.bitcoin.NativeSecp256k1.secp256k1_ec_pubkey_decompress
  5.8%     0  +    23    java.lang.Object.clone
  4.5%     0  +    18    java.lang.Throwable.fillInStackTrace
  3.3%     0  +    13    java.util.zip.Inflater.inflateBytes
  2.8%     0  +    11    sun.misc.Unsafe.copyMemory
  2.5%     0  +    10    java.lang.reflect.Array.newArray
  1.0%     3  +     1    java.lang.ClassLoader.defineClass1
  0.8%     0  +     3    java.lang.ClassLoader.findBootstrapClass
  0.8%     0  +     3    java.util.zip.ZipFile.getEntry
  0.3%     0  +     1    java.util.zip.Inflater.reset
  0.3%     0  +     1    java.util.zip.ZipFile.read
 56.9%     3  +   223    Total stub

  Thread-local ticks:
  0.3%     1             Class loader


Global summary of 4.37 seconds:
100.0%   408             Received ticks
  2.2%     9             Received GC ticks
135.0%   551             Compilation
  0.2%     1             Class loader
```

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala [hunk 1]
```diff
@@ -14,34 +14,35 @@ import scodec.bits.ByteVector
...
+sealed abstract class P2WPKHWitnessV0 extends ScriptWitnessV0 {
   def pubKey: ECPublicKey = ECPublicKey(stack.head)
 
   def signature: ECDigitalSignature = stack(1) match {
     case ByteVector.empty => EmptyDigitalSignature
     case bytes: ByteVector => ECDigitalSignature(bytes)
   }
 
-  override def toString = "P2WPKHWitnessV0(" + stack.map(BitcoinSUtil.encodeHex(_)).toString + ")"
+  override def toString = s"P2WPKHWitnessV0(${stack.map(BitcoinSUtil.encodeHex(_)).toString})"
```
Comment (torkelrogstad) [inline core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala:45]: Is this just for readability, or are there performance issues here?

File: core/src/main/scala/org/bitcoins/core/script/arithmetic/ArithmeticOperations.scala [hunk 2]
```diff
@@ -4,148 +4,148 @@ import org.bitcoins.core.script.ScriptOperationFactory
 import org.bitcoins.core.script.constant.ScriptOperation
 
 /** Created by chris on 1/6/16. */
-sealed trait ArithmeticOperation extends ScriptOperation
+sealed abstract class ArithmeticOperation extends ScriptOperation
```
Comment (torkelrogstad) [inline core/src/main/scala/org/bitcoins/core/script/arithmetic/ArithmeticOperations.scala:7]: This is for Java interop?

**Comment by torkelrogstad** [comment]
Very nice! Do you have any idea if it is the `val`/`def`s that did it, or the removal of the serializer loop?

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala (hunk 1 above)
Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala:45]: technically there are performance issues on multiple fronts

1. concatenating strings are generally expensive on the jvm [source](https://stackoverflow.com/questions/10078912/best-practices-performance-mixing-stringbuilder-append-with-string-concat)

2. i should really be caching the val 
```
stack.map(BitcoinSUtil.encodeHex(_)).toString
```

File: core/src/main/scala/org/bitcoins/core/script/arithmetic/ArithmeticOperations.scala (hunk 2 above)
Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/script/arithmetic/ArithmeticOperations.scala:7]: Yes -- for the `scalac` 2.11.x -- but apparently for `scalac`  2.12.x this doesn't matter any more https://www.scala-lang.org/news/2.12.0/

**Comment by Christewart** [comment]
@torkelrogstad It was mostly the changing from `def` -> `val`. This was a surprising performance increase but i guess it makes sense since by definition `case object`s should never change -- so we should be using `val`s that don't have to get re-evaluated every single time it is executed. 

The removal of the loop was a notable performance increase, but secondary to `def` -> `val`

**Comment by torkelrogstad** [comment]
Interestring! And very impressive gains.

**Pull request merged**